File: lib/migration.ts
import { FALLBACK_PROJECT, LEGACY_SUBJECT_MAP, PROJECTS } from '@/constants';
import { Priority, RepeatRule, Task, TaskStatus } from '@/types';

/**
 * Migration des données persistées.
 *
 * L'ancien modèle recopiait le *nom* et la *couleur* de la matière dans chaque
 * tâche. Le modèle actuel ne conserve que `subjectId`. Cette fonction convertit
 * les enregistrements existants au chargement pour ne pas perdre les données
 * d'un utilisateur ayant déjà installé l'application.
 */

interface LegacyTask {
  subject?: string;
  subjectColor?: string;
  attachments?: unknown;
  [key: string]: unknown;
}

function normalizeProjectId(id: string): string {
  return LEGACY_SUBJECT_MAP[id] ?? id;
}

const LEGACY_NAME_MAP: Record<string, string> = {
  mathématiques: 'work',
  physique: 'work',
  histoire: 'personal',
  langues: 'personal',
  autre: 'inbox',
};

function projectIdFromName(name: string | undefined): string {
  if (!name) return FALLBACK_PROJECT.id;
  const normalized = name.trim().toLowerCase();
  if (LEGACY_NAME_MAP[normalized]) return LEGACY_NAME_MAP[normalized];
  const match = PROJECTS.find(
    (p) => p.id === name || p.name.toLowerCase() === normalized
  );
  return match ? match.id : FALLBACK_PROJECT.id;
}

function isNonEmptyString(value: unknown): value is string {
  return typeof value === 'string' && value.length > 0;
}

/** Normalise une entrée inconnue ; renvoie `null` si elle est inexploitable. */
export function migrateTask(raw: unknown): Task | null {
  if (typeof raw !== 'object' || raw === null) return null;
  const input = raw as LegacyTask;

  if (!isNonEmptyString(input.id) || !isNonEmptyString(input.title)) return null;

  const rawProjectId = isNonEmptyString(input.subjectId)
    ? input.subjectId
    : projectIdFromName(input.subject);
  const subjectId = normalizeProjectId(rawProjectId);

  return {
    id: input.id,
    title: input.title,
    description: typeof input.description === 'string' ? input.description : '',
    dueDate: isNonEmptyString(input.dueDate) ? input.dueDate : new Date().toISOString(),
    priority: (['low', 'medium', 'high'] as Priority[]).includes(input.priority as Priority)
      ? (input.priority as Priority)
      : 'medium',
    subjectId,
    status: (['active', 'completed', 'archived'] as TaskStatus[]).includes(
      input.status as TaskStatus
    )
      ? (input.status as TaskStatus)
      : 'active',
    subtasks: Array.isArray(input.subtasks) ? (input.subtasks as Task['subtasks']) : [],
    reminders: Array.isArray(input.reminders) ? (input.reminders as string[]) : [],
    reportedAt: isNonEmptyString(input.reportedAt) ? input.reportedAt : null,
    durationMinutes: typeof input.durationMinutes === 'number' ? input.durationMinutes : null,
    note: typeof input.note === 'string' ? input.note : '',
    repeat: (['none', 'daily', 'weekly', 'monthly'] as RepeatRule[]).includes(
      input.repeat as RepeatRule
    )
      ? (input.repeat as RepeatRule)
      : 'none',
    createdAt: isNonEmptyString(input.createdAt) ? input.createdAt : new Date().toISOString(),
    completedAt: isNonEmptyString(input.completedAt) ? input.completedAt : null,
    archivedAt: isNonEmptyString(input.archivedAt) ? input.archivedAt : null,
  };
}

export function migrateTasks(raw: unknown): Task[] {
  if (!Array.isArray(raw)) return [];
  return raw.map(migrateTask).filter((t): t is Task => t !== null);
}

File: flutter/lib/src/logic/migration.dart
import '../constants.dart';
import '../models/task.dart';

/// Migration des données persistées, portée de `lib/migration.ts`.
///
/// L'ancien modèle recopiait le *nom* et la *couleur* de la matière dans chaque
/// tâche ; le modèle actuel ne conserve que `subjectId`. Ces fonctions
/// convertissent les enregistrements existants au chargement pour ne rien perdre
/// chez un utilisateur qui avait déjà installé l'application.
library;

const Map<String, String> _legacyNameMap = <String, String>{
  'mathématiques': 'work',
  'physique': 'work',
  'histoire': 'personal',
  'langues': 'personal',
  'autre': 'inbox',
};

String _normalizeProjectId(String id) => kLegacySubjectMap[id] ?? id;

String _projectIdFromName(Object? name) {
  if (name is! String || name.isEmpty) return kFallbackProject.id;
  final normalized = name.trim().toLowerCase();
  final byLegacyName = _legacyNameMap[normalized];
  if (byLegacyName != null) return byLegacyName;
  for (final project in kProjects) {
    if (project.id == name || project.name.toLowerCase() == normalized) return project.id;
  }
  return kFallbackProject.id;
}

bool _isNonEmptyString(Object? value) => value is String && value.isNotEmpty;

String? _stringOrNull(Map<String, Object?> input, String key) {
  final value = input[key];
  return _isNonEmptyString(value) ? value as String : null;
}

/// Normalise une entrée inconnue ; renvoie `null` si elle est inexploitable.
Task? migrateTask(Object? raw) {
  if (raw is! Map<dynamic, dynamic>) return null;
  final input = raw.cast<String, Object?>();

  final id = _stringOrNull(input, 'id');
  final title = _stringOrNull(input, 'title');
  if (id == null || title == null) return null;

  final storedSubjectId = _stringOrNull(input, 'subjectId');
  final subjectId = _normalizeProjectId(storedSubjectId ?? _projectIdFromName(input['subject']));

  final subtasks = input['subtasks'];
  final reminders = input['reminders'];
  final duration = input['durationMinutes'];

  return Task(
    id: id,
    title: title,
    description: (input['description'] as String?) ?? '',
    dueDate: _stringOrNull(input, 'dueDate') ?? DateTime.now().toUtc().toIso8601String(),
    priority: Priority.fromWire(input['priority']),
    subjectId: subjectId,
    status: TaskStatus.fromWire(input['status']),
    subtasks: subtasks is List<dynamic>
        ? subtasks
            .whereType<Map<dynamic, dynamic>>()
            .map((raw) => Subtask.fromJson(raw.cast<String, Object?>()))
            .toList(growable: false)
        : const <Subtask>[],
    reminders: reminders is List<dynamic>
        ? reminders.whereType<String>().toList(growable: false)
        : const <String>[],
    reportedAt: _stringOrNull(input, 'reportedAt'),
    durationMinutes: duration is num ? duration.toInt() : null,
    note: (input['note'] as String?) ?? '',
    repeat: RepeatRule.fromWire(input['repeat']),
    createdAt: _stringOrNull(input, 'createdAt') ?? DateTime.now().toUtc().toIso8601String(),
    completedAt: _stringOrNull(input, 'completedAt'),
    archivedAt: _stringOrNull(input, 'archivedAt'),
  );
}

/// Migre une liste brute ; un stockage corrompu donne une liste vide.
List<Task> migrateTasks(Object? raw) {
  if (raw is! List<dynamic>) return <Task>[];
  return raw.map(migrateTask).whereType<Task>().toList(growable: false);
}
